server/server.go

```diff
@@ -721,6 +739,16 @@ func (j *jsonRPCHub) GetCode(root types.Hash, addr types.Address) ([]byte, error
 	return code, nil
 }
 
+// Has returns true if the DB does contains the given key.
+func (j *jsonRPCHub) Has(rootHash types.Hash) bool {
+	snap, err := j.state.NewSnapshotAt(rootHash)
```

Comment:
IMO we should extend State interface with Has method, that way we will avoid snapshot getting. You have already added Has into immutable-trie/state.go, just extend State interface with Has and call state.Has directly from here.